File: src/native/strsim.ts
/**
 * String similarity — TypeScript boundary over the C wasm module.
 *
 * Powers "did you mean ...?" suggestions for unknown Telegram slash
 * commands and unknown CLI subcommands. The C source lives in
 * native/strsim-c and exports a three-function C ABI (alloc / dealloc /
 * levenshtein) — see that module's README for the contract, and
 * src/native/runtime.ts for the embedding and memory conventions
 * shared by every native module.
 *
 * Distance is computed over UTF-8 bytes. For the ASCII identifiers
 * this module exists for (command names, subcommands, model ids) byte
 * distance equals character distance; non-ASCII text weighs a
 * substituted codepoint as up to 4 edits, which still ranks fine.
 *
 * Strings longer than 1024 bytes come back as the wasm overflow
 * sentinel: `levenshtein` throws, `closestMatch` skips the candidate.
 */

import { STRSIM_WASM_BASE64 } from "./strsim-wasm-bytes.js";
import {
  allocRegion,
  embeddedWasm,
  toBytes,
  writeRegion,
  type WasmCoreExports,
} from "./runtime.js";

/** Returned by the wasm side when either input exceeds its DP buffer. */
const OVERFLOW = 0xffffffff;

/** The C-ABI surface exported by native/strsim-c. */
interface StrsimExports extends WasmCoreExports {
  levenshtein(aPtr: number, aLen: number, bPtr: number, bLen: number): number;
}

const strsimWasm = embeddedWasm<StrsimExports>(STRSIM_WASM_BASE64);

function rawDistance(a: Uint8Array, b: Uint8Array): number {
  const wasm = strsimWasm.instance();
  // Allocate both regions BEFORE writing either (runtime.ts convention:
  // alloc may grow linear memory and detach views).
  const aPtr = allocRegion(wasm, a.length, "levenshtein");
  let bPtr = 0;
  try {
    bPtr = allocRegion(wasm, b.length, "levenshtein");
    writeRegion(wasm, aPtr, a);
    writeRegion(wasm, bPtr, b);
    // wasm i32 return crosses as signed; >>> 0 restores the sentinel.
    return wasm.levenshtein(aPtr, a.length, bPtr, b.length) >>> 0;
  } finally {
    wasm.dealloc(bPtr, b.length);
    wasm.dealloc(aPtr, a.length);
  }
}

/**
 * Levenshtein edit distance between two strings (UTF-8 byte edits).
 * Throws for inputs longer than the wasm module's 1024-byte ceiling —
 * identifiers, not documents.
 */
export function levenshtein(a: string, b: string): number {
  const distance = rawDistance(toBytes(a), toBytes(b));
  if (distance === OVERFLOW) {
    throw new Error("levenshtein: input exceeds 1024 bytes");
  }
  return distance;
}

export interface ClosestMatch {
  value: string;
  distance: number;
}

/**
 * The candidate closest to `input` by edit distance, or null when none
 * comes within `maxDistance`. Ties keep the first candidate. Matching
 * is case-insensitive (identifiers like /Doctor vs /doctor); the
 * returned `value` keeps the candidate's original casing. Candidates
 * (or an input) beyond the 1024-byte ceiling are skipped, not errors.
 */
export function closestMatch(
  input: string,
  candidates: readonly string[],
  maxDistance = 2,
): ClosestMatch | null {
  const inputBytes = toBytes(input.toLowerCase());
  let best: ClosestMatch | null = null;
  for (const candidate of candidates) {
    const distance = rawDistance(inputBytes, toBytes(candidate.toLowerCase()));
    if (distance === OVERFLOW) continue;
    if (distance <= maxDistance && (!best || distance < best.distance)) {
      best = { value: candidate, distance };
      if (distance === 0) break;
    }
  }
  return best;
}

File: native/strsim-c/src/strsim.c
/*
 * strsim — string-similarity core in C, compiled to wasm32-freestanding.
 *
 * Powers "did you mean ...?" suggestions for unknown Telegram slash
 * commands and unknown CLI subcommands. The TypeScript consumer is
 * src/native/strsim.ts; the ABI is alloc / dealloc (native/shared/
 * walloc.h) plus `levenshtein` over linear memory — see README.md for
 * the contract.
 *
 * Distance is computed over bytes. Inputs are UTF-8, and for the ASCII
 * identifiers this module exists for (command names, subcommands,
 * model ids) byte distance equals character distance. Non-ASCII text
 * weighs a substituted codepoint as up to 4 edits — acceptable for
 * ranking, documented at the wrapper.
 *
 * The DP uses one statically-allocated row (two-row trick folded into
 * one row + a diagonal carry), so `levenshtein` itself never touches
 * the allocator. Strings longer than STRSIM_MAX_LEN return the
 * STRSIM_OVERFLOW sentinel instead of a distance — the wrapper maps
 * that to "no match" / a thrown error.
 */

#include <stddef.h>
#include <stdint.h>

#include "../../shared/walloc.h"

#define STRSIM_MAX_LEN 1024u
#define STRSIM_OVERFLOW 0xFFFFFFFFu

static uint32_t row[STRSIM_MAX_LEN + 1u];

__attribute__((export_name("levenshtein"))) uint32_t levenshtein(
    const uint8_t *a, size_t a_len, const uint8_t *b, size_t b_len) {
  if (a_len > STRSIM_MAX_LEN || b_len > STRSIM_MAX_LEN) {
    return STRSIM_OVERFLOW;
  }
  if (a_len == 0) return (uint32_t)b_len;
  if (b_len == 0) return (uint32_t)a_len;

  for (size_t j = 0; j <= b_len; j++) {
    row[j] = (uint32_t)j;
  }
  for (size_t i = 1; i <= a_len; i++) {
    uint32_t diag = row[0]; /* D(i-1, j-1) */
    row[0] = (uint32_t)i;
    for (size_t j = 1; j <= b_len; j++) {
      uint32_t above = row[j]; /* D(i-1, j) */
      uint32_t best = diag + (a[i - 1u] == b[j - 1u] ? 0u : 1u);
      uint32_t del = above + 1u;
      uint32_t ins = row[j - 1u] + 1u;
      if (del < best) best = del;
      if (ins < best) best = ins;
      row[j] = best;
      diag = above;
    }
  }
  return row[b_len];
}
